activerecord/CHANGELOG.md

```diff
@@ -1,3 +1,25 @@
+*   Add option to merge association scopes to subqueries.
+
+    It removes a breaking change. Conditions that were not applied in
+    previous versions would be applied now giving people different results.
+
+    Set the option on an association:
+
+    ```ruby
+    class Author < ApplicationRecord
+      has_many :welcome_posts, -> { where(title: "welcome") } , class_name: "Post", apply_on_subqueries: true
+    end
+    ```
+
+    Then the association scope will be merged into the subquery.
+
+    ```ruby
+    Author.where(welcome_posts: Post.all)
```

Comment:
Even as an optional feature _and_ one which explicitly notes it only applies to subqueries, I'm really nervous about the way this departs from the non-subquery equivalent behaviour:

```ruby
Author.where(welcome_posts: Post.all.to_a)

Author.where(welcome_posts: Post.last)

# and yes, even
Author.where(welcome_posts: 1)
```

Perhaps it should actually be more like `apply_scope_on_conditions`, and while we can possibly[^1] use merge when the condition is a relation, we should double-barrel the SQL condition otherwise: `id IN (1, 2, 3) AND id IN (SELECT .. welcome)`

[^1]: I'm always nervous about `merge`; it has surprising (but locked in for compatibility) behaviour when both relations have overlapping-but-distinct opinions about the same clauses.

Reply:
Perhaps "scope" is too specific after all:

```ruby
Author.where(special_posts: Post.all)
```